### PR TITLE
Session Restore: Move to pre-section routing

### PR DESCRIPTION
**Pull request by mcsf**
If successful, supersedes #16291. Main advantages of this approach:
- If there is a session to restore, we skip loading the Reader module altogether. The visual glitch is also gone, and performance is obviously better.
- The Reader module is left untouched and is unaware of this mechanism.
- No virtual Calypso section needs to be cooked up for this.

This PR _cheats_ by using localStorage instead of localforage (Session Restore's MVP used localStorage until #14196).

See commit messages for more insight.

**Question:** localforage is asynchronous, so to use it we'd have to refactor `server/bundler/loader` accordingly, which is something I would consider avoiding. Maybe `localStorage` could be good enough for this?

**Contributing:** If you are unfamiliar with `server/bundler/loader`, know that every change there will require a full Calypso rebuild to effect.

**Figure 1:** With no saved path to restore

**Figure 2:** With a saved path to restore

## TO TEST

**NOTE:** Feel free to substitute http://calypso.localhost:3000/ for the following if you prefer to run this branch locally

* Start here: https://calypso.live/?branch=fix/path-restore-earlier
* Click around -- go wild!
  * Make sure you can browse to the Reader as before
* Close the tab (and / or open an additional one)
* Start back at https://calypso.live
* Note where you land -- is it expected & delightful?
* Repeat the above, but end your "session" at `/` (the Reader root)
  * Do you land at the `/` Reader when you return?
* Clear your localStorage (`localStorage.clear()` in the console or log out and back in)
* Browse to https://calypso.live
  * Do you land at the `/` Reader as expected?
* Try to break stuff!

### Test with the config flag disabled:

* Run locally with the `restore-last-location` flag disabled & the `no-force-sympathy` forced on to make sure your local storage doesn't get cleared out:
  `ENABLE_FEATURES=no-force-sympathy DISABLE_FEATURES=restore-last-location npm start`
* Browse to http://calypso.localhost:3000/
* Whitelist the debugger scope in your console:
  * `localStorage.setItem( 'debug', 'calypso:restore-last-location' )`
* Click around
  * You should not see debug messages related to this feature
* Browse somewhere other than `/` & close the tab
* Browse to http://calypso.localhost:3000/
  * You should land at the `/` Reader (i.e. this feature is decidedly not active)

## TODO before merging
- [x] Update the [middleware README](https://github.com/Automattic/wp-calypso/blob/578c6b776330658ed6948cc8eb0732bfe26d8568/client/state/routing/README.md) to indicate the change in code structure here
- [x] Add a README to the [restore-last-path lib](https://github.com/Automattic/wp-calypso/tree/578c6b776330658ed6948cc8eb0732bfe26d8568/client/lib/restore-last-path)

### CODE REVIEW

**Comment by mcsf** [comment]
Added a couple of GIFs because we love them.

**Comment by jblz** [comment]
Rebased to try and fix CI failures

**Comment by mcsf** [comment]
PR description includes the following:

> **Question:** localforage is asynchronous, so to use it we'd have to refactor `server/bundler/loader` accordingly, which is something I would consider avoiding. Maybe `localStorage` could be good enough for this?

Which can be of interest to @gwwar. :-) (N.B.: I will be AFK for the next couple of days)

**Comment by gwwar** [comment]
That's fine, but just note the reasons why you're choosing localStorage here as a comment or in a readme.

**Comment by lancewillett** [comment]
Tested the Calypso Live Branches flows as described above in Safari, Chrome, and Firefox on macOS. Nice! Everything works as expected when I used My Sites or Reader URLs as the session "ending point."

Is there a whitelist in place? I noticed in testing `/me` URLs as the exit point, they weren't restored, and I was taken back to my previously saved "whitelisted" URL under My Sites.

**Comment by jblz** [comment]
@lancewillett yes, there's a whitelist [here](https://github.com/Automattic/wp-calypso/blob/dd685ad265e967abcc66a9d26ba9d53f3d5dadaa/client/lib/restore-last-path/index.js#L12).

Thanks for testing!

File: client/lib/restore-last-path/index.js [hunk 1]
```diff
@@ -35,37 +48,73 @@ function validatePath( path ) {
...
 	} );
 }
 
-export default {
-	isWhitelistedForRestoring,
-	getSavedPath,
-	savePath,
-};
+// Store a module-level variable to enforce restoreLastSession only runs once per session
+let isFirstRun = true;
```
Comment (jblz) [inline client/lib/restore-last-path/index.js]: I like it -- will give it a spin

Comment (marekhrabe) [inline client/lib/restore-last-path/index.js]: We expect to get a boolean from this function. What will `once` do? Return undefined?

Comment (marekhrabe) [inline client/lib/restore-last-path/index.js]: `once` will return the first returned value for all subsequent calls. If the path was initially restored, this function will keep returning `true` forever. https://lodash.com/docs/4.17.4#once

Comment (jblz) [inline client/lib/restore-last-path/index.js]: Yep, I'm afraid that wont work...

We'd kind of need a `onceThenAlwaysFalse` or `onceThenAlways` function.

Comment (spen) [inline client/lib/restore-last-path/index.js]: Yeah, I'd wrote something to that effect but it didn't send :/ I'm keen to explore it a little more over the weekend but definitely happy to ignore it for now :)

File: client/state/routing/middleware.js [hunk 2]
```diff
@@ -3,50 +3,28 @@
...
+import { savePath } from 'lib/restore-last-path';
 import { ROUTE_SET } from 'state/action-types';
 
 const debug = debugFactory( 'calypso:restore-last-location' );
-let hasInitialized = false;
 
 export const routingMiddleware = () => {
 	return ( next ) => ( action ) => {
-		if ( action.type !== ROUTE_SET || ! action.path || ! action.query ) {
+		if ( action.type !== ROUTE_SET ||
```
Comment (spen) [inline client/state/routing/middleware.js]: completely ignorable, personal-opinion-based nit: 
What do you think to a format like so?:
```js
if (
    x === y,
    a !== b
) {
   return c;
}
```
My only rational for this is for slightly clearer indentation and boundaries between condition and logic - so FFTI.

File: client/state/routing/middleware.js [hunk 3]
```diff
@@ -3,50 +3,28 @@
...
+				! action.path ||
+				! action.query ||
```
Comment (spen) [inline client/state/routing/middleware.js]: I think we may be able to ommit `! action.query ||` here as that condition should be covered by `isEmpty`

From the [lodash REPL](https://lodash.com/docs/4.17.4#isEmpty)
```js
a = {
    b: [],
    c: [1]
}

_.isEmpty( a.d );
// => true
```

**Comment by jblz** [comment]
Thanks for the suggestions, @spen -- incorporated both.

**Pull request merged**